### PR TITLE
Remove reflection-based deserialization setters

### LINKED ISSUE #16353: Repository Quality: Reflection Anti-patterns and Private-Setter Bypass Audit
### 🎯 Repository Quality Improvement Report — Reflection Anti-patterns and Private-Setter Bypass Audit

**Analysis Date**: 2026-08-12
**Focus Area**: reflection-anti-patterns-private-setter-bypass
**Strategy Type**: Custom

### Executive Summary

An audit of the 46 source files that use reflection reveals a cluster of anti-patterns centred on **bypassing private setters during deserialization** and **late-binding to external assemblies via string-based method lookup**. The most critical findings:

- `TestSessionInfo.Id` has `private set`; two separate converters use `GetProperty(...).SetValue(...)` reflection to bypass it — incompatible with NativeAOT/trimming.
- `DiscoveryCriteria` has four `private set` properties bypassed in both the legacy Jsonite path and System.Text.Json path via reflection. The Jsonite fallback also uses the compiler-internal `<{prop.Name}>k__BackingField` naming convention.
- `CodeCoverageDataAttachmentsHandler.cs` retrieves `Task.Result` via `GetProperty("Result").GetValue(...)` when a direct cast to `Task<IList<string>>` is sufficient.

<details>
<summary><b>Full Analysis</b></summary>

### Metrics

| Metric | Value | Status |
|--------|-------|--------|
| Source files using reflection APIs | 46 | ⚠️ |
| `Activator.CreateInstance` call sites | 16 | ⚠️ |
| `Type.GetType` call sites | 6 | ⚠️ |
| Private-setter bypasses via `SetValue` | 6 across 4 converters | ❌ |
| Compiler-internal `k__BackingField` hacks | 1 | ❌ |
| `Task.Result` accessed via reflection | 1 | ❌ |

### Key Findings

**[Critical] TestSessionInfo.Id private-setter bypass — duplicated across two converters**
`JsoniteConvert.cs:409` and `TestSessionInfoConverter.cs:36` both execute:
```csharp
typeof(TestSessionInfo).GetProperty(nameof(TestSessionInfo.Id))!.SetValue(info, id);
```
Root cause: `TestSessionInfo` has no constructor that accepts a known `Guid`.

**[Critical] DiscoveryCriteria private-setter bypass in two serialization paths**
`JsoniteConvert.cs:417-420` and `DiscoveryCriteriaConverter.cs:45-48` both use `GetProperty(...).SetValue(...)` for `AdapterSourceMap`, `FrequencyOfDiscoveredTestsEvent`, `DiscoveredTestEventTimeout`, and `RunSettings`. The Jsonite path also falls back to writing directly into `<{prop.Name}>k__BackingField` — an undocumented compiler implementation detail.

**[High] Task.Result via reflection in CodeCoverageDataAttachmentsHandler.cs**
`CodeCoverageDataAttachmentsHandler.cs:135`: `task.GetType().GetProperty("Result")!.GetValue(task, null)` — a direct cast to `Task<IList<string>>` eliminates the reflection call.

**[Medium] Null-safety gap in TestExtensions.cs**
`Type.GetType(testPluginInformation?.AssemblyQualifiedName!)` — the `!` suppresses the warning but `AssemblyQualifiedName` can still be null at runtime; `Type.GetType(null)` returns null silently.

</details>

---

#### Task 1: Add `internal` deserialization constructor to `TestSessionInfo`

**Priority**: High | **Effort**: Small

Add `internal TestSessionInfo(Guid id) { Id = id; }` to `src/Microsoft.TestPlatform.ObjectModel/Client/TestSessionInfo.cs`, then replace `GetProperty(...).SetValue(...)` in both:
- `src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/JsoniteConvert.cs:408-409`
- `src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/TestSessionInfoConverter.cs:35-36`

---

#### Task 2: Expose `internal set` on `DiscoveryCriteria` private-setter properties

**Priority**: High | **Effort**: Medium

Change `private set` to `internal set` on `AdapterSourceMap`, `FrequencyOfDiscoveredTestsEvent`, `DiscoveredTestEventTimeout`, and `RunSettings` in `src/Microsoft.TestPlatform.ObjectModel/Client/DiscoveryCriteria.cs`. Update both converters to assign directly rather than via `GetProperty(...).SetValue(...)`.

---

#### Task 3: Remove the `k__BackingField` compiler-internal hack from `JsoniteConvert.cs`

**Priority**: Medium | **Effort**: Small

After Tasks 1 and 2 land, verify no other serialized type depends on the backing-field fallback (`JsoniteConvert.cs:563`) and remove it. Replace with an `EqtTrace.Warning` when no accessible setter exists, making failures visible.

---

#### Task 4: Replace `Task.Result` reflection in `CodeCoverageDataAttachmentsHandler.cs`

**Priority**: Medium | **Effort**: Small

Replace `task.GetType().GetProperty("Result")!.GetValue(task, null)` (`CodeCoverageDataAttachmentsHandler.cs:135`) with a direct cast to `Task<IList<string>>`. Also add a check that `s_mergeMethodInfo.ReturnType` is assignable to `Task<IList<string>>` when caching the `MethodInfo` to fail fast at initialization rather than at call time.

---

#### Task 5: Fix null-safety gap in `TestExtensions.cs` and document `throwOnError` in `TestPluginManager.cs`

**Priority**: Low | **Effort**: Small

- `TestExtensions.cs:431`: replace `testPluginInformation?.AssemblyQualifiedName!` with an explicit null guard before passing to `Type.GetType`.
- `TestPluginManager.cs:36`: wrap `Type.GetType(extensionTypeName, true)` in a targeted `catch (TypeLoadException)` that logs the type name at error level, so the failure is not swallowed by the broad `catch (Exception)` in callers.

---

### 📊 Historical Context

<details>
<summary><b>Previous Focus Areas (last 10)</b></summary>

| Date | Focus Area | Type |
|------|------------|------|
| 2026-08-11 | conditional-compilation-dead-code-and-legacy-symbol-hygiene | Custom |
| 2026-08-10 | permanently-ignored-test-coverage-debt | Custom |
| 2026-08-07 | timeout-budget-architecture-and-consistency | Custom |
| 2026-08-06 | assembly-architecture-inference-coverage | Custom |
| 2026-08-05 | environment-variable-propagation-testhost-setup | Custom |
| 2026-08-04 | test-case-filter-engine-correctness-and-edge-cases | Custom |
| 2026-08-03 | telemetry-opt-in-correctness-and-data-hygiene | Custom |
| 2026-07-31 | testproperty-propertybag-type-safety-and-serialization-correctness | Custom |
| 2026-07-30 | logger-concurrency-correctness-and-output-reliability | Custom |
| 2026-07-29 | ipc-connection-resilience-timeout-arithmetic-correctness | Custom |

</details>

*Next analysis: 2026-08-13 — Focus area selected based on diversity algorithm*

> Generated by [Repository Quality Improver](https://github.com/microsoft/vstest/actions/runs/31562886354) · 62.6 AIC · ⌖ 8.48 AIC · ⊞ 8.9K · [◷](https://github.com/search?q=repo%3Amicrosoft%2Fvstest+is%3Aissue+%22gh-aw-workflow-call-id%3A+microsoft%2Fvstest%2Frepository-quality-improver%22&type=issues)
> - [x] expires  on Aug 14, 2026, 4:26 AM UTC

### PR DESCRIPTION
**Pull request by nohwnd**
Reflection was used to write private ObjectModel setters in both serializers, and Jsonite also depended on compiler-generated backing-field names.

This gives CommunicationUtilities internal access to the deserialization state and assigns it directly. Jsonite reconstructs immutable data-collection IDs through their constructors instead of writing backing fields. The optional code coverage loader still uses reflection, but awaits its declared task result directly. Invalid plugin metadata is now diagnosed and skipped.

Fix #16353

## Verification

- `build.cmd -c Release`
- CommunicationUtilities unit tests: 584/584 on net11.0, 585/585 on net481
- Utilities unit tests: 123/123 on net11.0 and net481
- Common unit tests: 394 passed, 2 existing skips on net11.0 and net481
- NativeAOT converter publish and runtime round trip: 1/1

🤖

### CODE REVIEW

File: src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/DiscoveryCriteriaConverter.cs [hunk 1]
```diff
@@ -38,14 +37,13 @@ internal class DiscoveryCriteriaConverter : JsonConverter<DiscoveryCriteria>
...
-        type.GetProperty(nameof(DiscoveryCriteria.FrequencyOfDiscoveredTestsEvent))!.SetValue(criteria, frequency);
-        type.GetProperty(nameof(DiscoveryCriteria.DiscoveredTestEventTimeout))!.SetValue(criteria, timeout);
-        type.GetProperty(nameof(DiscoveryCriteria.RunSettings))!.SetValue(criteria, runSettings);
+        var criteria = new DiscoveryCriteria
+        {
+            AdapterSourceMap = adapterSourceMap!,
+            FrequencyOfDiscoveredTestsEvent = frequency,
+            DiscoveredTestEventTimeout = timeout,
+            RunSettings = runSettings,
+        };
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply src/Microsoft.TestPlatform.CommunicationUtilities/Serialization/DiscoveryCriteriaConverter.cs]: Fixed in 3078ace953. `DeserializeProperty` now falls back to an empty dictionary, so a payload that omits `AdapterSourceMap` or sets it to null deserializes into a usable `DiscoveryCriteria` instead of throwing.

Worth noting the .NET Framework path in `JsoniteConvert.DeserializeDiscoveryCriteria` already defaulted to `[]` here, so the two serializers were disagreeing — they now behave the same.

Added `DeserializePayloadWithoutAdapterSourceMapProducesEmptyAdapterMap` and `DeserializePayloadWithNullAdapterSourceMapProducesEmptyAdapterMap`; both throw `NullReferenceException` without the change.

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/NativeAotCompatibilityTests.cs [hunk 2]
```diff
@@ -93,12 +102,56 @@ public void TranslationLayer_NativeAotPublish_ShouldNotProduceSerializationWarni
...
+            "..",
+            "artifacts",
+            "bin",
+            "TestAssets",
+            "NativeAotTranslationLayerConsumer",
+            "Release",
+            "net8.0",
+            rid,
+            "publish",
+            executableName);
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/NativeAotCompatibilityTests.cs]: The configuration mismatch was fixed in f7af2391d4 — `dotnet publish` is now invoked with `-c Release`, matching the hardcoded path.

Running it locally turned up a second problem on this same path, fixed in 3078ace953: the path was built with literal `..` segments and never normalized. `Process.Start` measures the path as given, so on a longer checkout path it went over MAX_PATH and failed with "The filename or extension is too long". CI did not hit it because its checkout path is short. It is now wrapped in `Path.GetFullPath`, with a `File.Exists` assertion so a future mismatch reports the path instead of a `Win32Exception`.

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: test/TestAssets/NativeAotTranslationLayerConsumer/NativeAotTranslationLayerConsumer.csproj [hunk 3]
```diff
@@ -12,12 +15,17 @@
     <TrimmerSingleWarn>false</TrimmerSingleWarn>
     <!-- Treat trimming/AoT warnings as errors so they fail the publish. -->
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <!-- The compatibility test evaluates linker/AOT warnings after publish so it can exclude
+         known warnings outside the serialization converters under test. -->
+    <ILLinkTreatWarningsAsErrors>false</ILLinkTreatWarningsAsErrors>
+    <IlcTreatWarningsAsErrors>false</IlcTreatWarningsAsErrors>
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply test/TestAssets/NativeAotTranslationLayerConsumer/NativeAotTranslationLayerConsumer.csproj]: Fixed the comment in 3078ace953. It now says what the properties actually do: compiler warnings fail the build, trimming/AoT warnings deliberately do not, because the publish has to succeed for the test to read the warning list and assert on the subset coming from the serialization converters.

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

File: test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/NativeAotCompatibilityTests.cs [hunk 4]
```diff
@@ -59,6 +59,15 @@ public void TranslationLayer_NativeAotPublish_ShouldNotProduceSerializationWarni
             CreateNoWindow = true,
         };
 
+        if (OperatingSystem.IsWindows())
+        {
+            var visualStudioInstallerPath = Path.Combine(
+                Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86),
+                "Microsoft Visual Studio",
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply test/Microsoft.TestPlatform.CommunicationUtilities.UnitTests/NativeAotCompatibilityTests.cs:66]: `dotnet publish` now runs with an explicit `-c Release` (f7af2391d4), so it matches the hardcoded output path.

3078ace953 also normalizes that path with `Path.GetFullPath`. The unnormalized form kept its `..` segments and `Process.Start` failed with "The filename or extension is too long" once the checkout path was long enough, even though the publish itself had succeeded.

> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

**Comment by nohwnd** [comment]
Answering the two suppressed comments from the latest review. Both are addressed in 67937249b2.

**`JsoniteConvert.CreateDataCollectionContext`** - correct, that reflection was pointless. `protected internal` means protected *or* internal, and this PR made CommunicationUtilities a friend assembly of ObjectModel, so the constructors are directly reachable. `Microsoft.TestPlatform.Common` already calls them that way. Both helpers are gone now, replaced by `new DataCollectionContext(sessionId, testCase)` and `new DataCollectionContext(sessionId, testExecId)`. `DeserializePayloadPreservesDataCollectionContextIds` covers that path.

**`CodeCoverageDataAttachmentsHandler` merge result** - the mechanism is right, `Task<T>` is invariant, so a declared `Task<List<string>>` would stop matching and we would return the files unmerged. It is not the case today though: `Microsoft.CodeCoverage.IO` 17.7.0 declares `Task<IList<string>>`, and the existing merge tests run against the real assembly, so this is proven, not assumed.

I did not add normalization for return shapes that do not exist, that is untested code for a hypothetical package version. Instead I made a future change visible:

- `MergeCoverageReportsAsyncReturnsTaskOfIListOfString` pins the declared return type. A package bump that changes it fails this test instead of silently degrading coverage merging.
- The error trace now names the method, the expected type and the type actually returned, instead of just "Failed to merge code coverage files".

Utilities unit tests 124/124 and CommunicationUtilities 614/614, on net11.0 and net481.

🤖

**Comment by nohwnd** [comment]
Both suppressed comments from the latest review make the same point, and it is a fair one: `Microsoft.TestPlatform.CommunicationUtilities` is a shipping assembly, and this PR was granting `InternalsVisibleTo` to a test asset signed with `TITestDllKey.snk`, which is checked into the repository. Anyone can sign an assembly with that key, so the grant was effectively open.

Fixed in b40a6b3bcd by taking the suggested route. The consumer now round trips through `JsonDataSerializer.Instance.SerializePayload` / `DeserializeMessage` / `DeserializePayload<T>`, the same public entry point vstest.console and testhost use on the wire. `JsonDataSerializer` registers `DiscoveryCriteriaConverter` and `TestSessionInfoConverter` and resolves through `TestPlatformJsonContext`, so the linker still walks the same code, and the consumer needs no internals at all.

Removed with it:

- the `InternalsVisibleTo("NativeAotTranslationLayerConsumer", ...)` entry in `src/Microsoft.TestPlatform.CommunicationUtilities/Friends.cs`
- `AssemblyOriginatorKeyFile`, `SignAssembly` and `PublicSign` from the consumer project, which only existed to satisfy that grant

This also makes the check closer to what a real consumer does, since it now exercises the public entry point rather than constructing internal converters by hand.

Verified by running the test itself: `TranslationLayer_NativeAotPublish_ShouldNotProduceSerializationWarnings` passes, so the unsigned consumer publishes with `PublishAot=true`, reports no serialization warnings from CommunicationUtilities, and the published binary runs the round trip. CommunicationUtilities 614/614 on net11.0 and net481.

🤖

**Review by 0101 — APPROVED** [review]

**Comment by nohwnd** [comment]
Answering the last suppressed comment, the one about `psi.Environment["PATH"]`. Leaving the code as it is, here is the reasoning.

The block is Windows only, and `psi.Environment` is seeded from the current process environment. For the indexer to throw, PATH would have to be missing from that environment, and in that case `Process.Start` has nothing to search when it resolves `FileName = "dotnet"`, because that lookup goes through PATH as well. `dotnet.exe` sits in `C:\Program Files\dotnet`, not in System32 or the Windows directory, so the publish step fails either way. `TryGetValue` would move the failure a few lines down and change which exception you see, the test still cannot pass.

The related problem that would be real is casing, and that one is already handled. Windows spells the variable `Path`, not `PATH`, but `ProcessStartInfo.Environment` is built with `StringComparer.OrdinalIgnoreCase` on Windows (and `Ordinal` elsewhere), so the indexer finds it whichever way it is spelled. Checked it rather than assumed it, `ContainsKey` returns true for `PATH`, `Path` and `path`.

Fair defensive coding point in general, it just does not describe a failure this test can reach, so same as with the merge result comment earlier, I did not add a branch that no test can enter.

🤖
